crates/yulgen/src/runtime/mod.rs

```diff
@@ -1,174 +1,248 @@
 pub mod abi_dispatcher;
 pub mod functions;
-use crate::types::{to_abi_types, AbiDecodeLocation, AbiType};
-use crate::Context;
-use fe_analyzer::context::FunctionAttributes;
-use fe_analyzer::namespace::types::Contract;
-use fe_parser::ast as fe;
-use fe_parser::node::Node;
+use crate::context::ContractContext;
+use crate::types::{to_abi_types, AbiDecodeLocation, AbiType, AsAbiType};
+use fe_analyzer::namespace::items::ContractId;
+use fe_analyzer::namespace::types::FunctionSignature;
+use fe_analyzer::AnalyzerDb;
+use std::rc::Rc;
 use yultsur::*;
 
 /// Builds the set of function statements that are needed during runtime.
-pub fn build(context: &Context, contract: &Node<fe::Contract>) -> Vec<yul::Statement> {
-    if let Some(attributes) = context.analysis.get_contract(contract) {
-        let std = functions::std();
-
-        let external_functions =
-            concat_contract_functions(attributes.external_contracts.to_owned());
-
-        let encoding = {
-            let public_functions_batch = attributes
-                .public_functions
-                .iter()
-                .filter(|attributes| !attributes.return_type.is_unit())
-                .map(|attributes| vec![AbiType::from(&attributes.return_type)])
-                .collect::<Vec<_>>();
-
-            let events_batch = attributes
-                .events
-                .iter()
-                .map(|event| to_abi_types(&event.non_indexed_field_types()))
-                .collect::<Vec<_>>();
-
-            let contracts_batch = external_functions
-                .iter()
-                .map(|function| to_abi_types(&function.param_types()))
-                .collect();
-
-            let assert_strings_batch = context
-                .assert_strings
-                .iter()
-                .map(|val| vec![AbiType::from(val)])
-                .collect::<Vec<_>>();
-
-            let revert_errors_batch = context
-                .revert_errors
+pub fn build(
+    db: &dyn AnalyzerDb,
+    context: &ContractContext,
+    contract: ContractId,
+) -> Vec<yul::Statement> {
+    // TODO: db should provide easier access to these things.
+    // Eg. contract.public_functions(db)
```

Comment:
Yeah, I fully agree. I was thinking the same thing as I was querying the functions map for "__init__", but I was desperate to just get stuff working and left it alone. I'll clean it up now though.

Reply:
@g-r-a-n-t see the latest commit. Calling `__init__` on self or another value of type contract is now disallowed (because it's not in the functions/public_functions maps, but also I added a specific check for it so the error message could be better). I didn't touch `Foo.__init__`; that's already an error and the nearby code would require some reshuffling to add a helpful note about create/create2, which I was too tired to do it (added a `// TODO:` though).

I also added an error if one tries to call a non-pub fn on a (non-self) contract.

Reply:
nice :+1: